File: ResourceManager/Services/UserDataService.cs
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Design;
using ResourceManager.Models;
using ResourceManager.Services.Abstractions;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace ResourceManager.Services
{
    public class UserDataService : GenericDataService<User>, IUserDataService
    {
        public UserDataService(IDesignTimeDbContextFactory<ResourceManagerDBContext> contextFactory) : base(contextFactory)
        {
        }
        public override async Task<IList<User>> GetAll()
        {
            using (ResourceManagerDBContext db = dbFactory.CreateDbContext(null!))
            {
                IList<User> entities = await db.Users.OrderBy(x => x.Name).ToListAsync();
                return entities;
            }
        }
        public async Task<int> UpdateRoles(User user, IList<Role> roles)
        {
            using (ResourceManagerDBContext db = dbFactory.CreateDbContext(null!))
            {
                var u = db.Users.Include(x => x.UserRoles).Single(u => u.ID == user.ID);
                db.UpdateManyToMany(u.UserRoles, roles.Select(x => new UserRole() { RoleID = x.ID, UserID = u.ID }), x=>x.RoleID);

                return await db.SaveChangesAsync();
            }
        }
    }
}

File: ResourceManager/Services/UserDataService.vb
Imports Microsoft.EntityFrameworkCore
Imports Microsoft.EntityFrameworkCore.Design
Imports ResourceManager.Models
Imports ResourceManager.Services.Abstractions

Namespace ResourceManager.Services
    Public Class UserDataService
        Inherits GenericDataService(Of User)
        Implements IUserDataService
        Public Sub New(ByVal contextFactory As IDesignTimeDbContextFactory(Of ResourceManagerDBContext))
            MyBase.New(contextFactory)
        End Sub
        Public Overrides Async Function GetAll() As Task(Of IList(Of User)) Implements IDataService(Of User).GetAll
            Dim db = MyBase.dbFactory.CreateDbContext(Nothing)
            Using db
                Dim entities = Await db.Users.OrderBy(Function(x) x.Name).ToListAsync()
                Return entities
            End Using
        End Function
        Public Async Function UpdateRoles(ByVal user As User, ByVal roles As IList(Of Role)) As Task(Of Integer) Implements IUserDataService.UpdateRoles
            Dim db = MyBase.dbFactory.CreateDbContext(Nothing)
            Using db
                Dim u = db.Users.Include(Function(x) x.UserRoles).Single(Function(x) x.ID = user.ID)
                db.UpdateManyToMany(u.UserRoles, roles.Select(Function(x) New UserRole With {.RoleID = x.ID, .UserID = u.ID}), Function(x) x.RoleID)
                Return Await db.SaveChangesAsync()
            End Using
        End Function
    End Class
End Namespace
